### PR TITLE
Add new filters to products

### PR DESCRIPTION
**Pull request by kadewu**
Add filters:
- isAvailable
- publishedFrom
- availableFrom
- isVisibleInListing

Fix [10370](https://github.com/saleor/saleor/issues/10370)

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/graphql/product/filters.py [hunk 1]
```diff
@@ -451,6 +452,46 @@ def _filter_products_is_published(qs, _, value, channel_slug):
...
+            Exists(channel.filter(pk=OuterRef("channel_id"))),
+            Q(available_for_purchase_at__gt=now)
+            | Q(available_for_purchase_at__isnull=True),
+        ).values("product_id")
+
+    return qs.filter(Exists(product_channel_listings.filter(product_id=OuterRef("pk"))))
+
+
+def _filter_products_channel_field_from_date(qs, _, value, channel_slug, field):
+    channel = Channel.objects.filter(slug=channel_slug).values("pk")
```
Comment (IKarbowiak) [inline saleor/graphql/product/filters.py:474]: You should handle the empty value here too. And please add test for this scenario

File: saleor/graphql/product/filters.py [hunk 2]
```diff
@@ -594,6 +635,24 @@ class ProductStockFilterInput(graphene.InputObjectType):
...
+    is_available = django_filters.BooleanFilter(
+        method="filter_is_available",
+        help_text=f"Filter by available for purchase. {ADDED_IN_38}",
+    )
+    available_from = ObjectTypeFilter(
+        input_class=graphene.DateTime,
+        method="filter_available_from",
+        help_text=f"Filter by available for purchase from date. {ADDED_IN_38}",
+    )
+    listed = django_filters.BooleanFilter(
```
[the remark this answers is not in the document]
Comment (kadewu) [reply saleor/graphql/product/filters.py:652]: I also wasn't sure about the name, so just followed the suggestion from issue. I like the proposal to change it though.

File: saleor/graphql/product/filters.py [hunk 3]
```diff
@@ -451,6 +452,46 @@ def _filter_products_is_published(qs, _, value, channel_slug):
     )
 
 
+def _filter_products_is_available(qs, _, value, channel_slug):
+    channel = Channel.objects.filter(slug=channel_slug).values("pk")
+    now = datetime.datetime.now(pytz.UTC)
+    if value:
```
[the remark this answers is not in the document]
Comment (kadewu) [reply saleor/graphql/product/filters.py:458]: I've just started applying changes, but after some thought I wonder if should I actually care about `None`. I mean, the `None` value is handled in `field`, so the `None` will never be passed here, and I've marked those methods as private. I can of course add that logic here for the future, but then I would need to introduce tests on 'private' method.

[the remark this answers is not in the document]
Comment (IKarbowiak) [reply saleor/graphql/product/filters.py:458]: Yep, you are right, I checked the call with providing `null` as boolean filter value and it didn't enter the method.

**Review by IKarbowiak — APPROVED** [review]

File: saleor/graphql/product/filters.py [hunk 4]
```diff
@@ -594,6 +635,24 @@ class ProductStockFilterInput(graphene.InputObjectType):
 
 class ProductFilter(MetadataFilterBase):
     is_published = django_filters.BooleanFilter(method="filter_is_published")
+    published_from = ObjectTypeFilter(
+        input_class=graphene.DateTime,
+        method="filter_published_from",
+        help_text=f"Filter by published from date. {ADDED_IN_38}",
```
Comment (maarcingebala) [inline saleor/graphql/product/filters.py:641]:
```suggestion
        help_text=f"Filter by the publication date. {ADDED_IN_38}",
```

File: saleor/graphql/product/filters.py [hunk 5]
```diff
@@ -594,6 +635,24 @@ class ProductStockFilterInput(graphene.InputObjectType):
...
+    )
+    is_available = django_filters.BooleanFilter(
+        method="filter_is_available",
+        help_text=f"Filter by available for purchase. {ADDED_IN_38}",
```
Comment (maarcingebala) [inline saleor/graphql/product/filters.py:645]:
```suggestion
        help_text=f"Filter by availability for purchase. {ADDED_IN_38}",
```

File: saleor/graphql/product/filters.py [hunk 6]
```diff
@@ -594,6 +635,24 @@ class ProductStockFilterInput(graphene.InputObjectType):
...
+    )
+    available_from = ObjectTypeFilter(
+        input_class=graphene.DateTime,
+        method="filter_available_from",
+        help_text=f"Filter by available for purchase from date. {ADDED_IN_38}",
+    )
+    is_visible_in_listing = django_filters.BooleanFilter(
+        method="filter_listed",
+        help_text=f"Filter by visible in product listings. {ADDED_IN_38}",
```
Comment (maarcingebala) [inline saleor/graphql/product/filters.py:654]:
```suggestion
        help_text=f"Filter by visibility in product listings. {ADDED_IN_38}",
```

File: saleor/graphql/product/filters.py [hunk 7]
```diff
@@ -594,6 +635,24 @@ class ProductStockFilterInput(graphene.InputObjectType):
...
+        help_text=f"Filter by published from date. {ADDED_IN_38}",
+    )
+    is_available = django_filters.BooleanFilter(
+        method="filter_is_available",
+        help_text=f"Filter by available for purchase. {ADDED_IN_38}",
+    )
+    available_from = ObjectTypeFilter(
+        input_class=graphene.DateTime,
+        method="filter_available_from",
+        help_text=f"Filter by available for purchase from date. {ADDED_IN_38}",
```
Comment (maarcingebala) [inline saleor/graphql/product/filters.py:650]:
```suggestion
        help_text=f"Filter by the date of availability for purchase. {ADDED_IN_38}",
```

**Review by maarcingebala — COMMENTED** [review]
I've redacted the descriptions slightly to make them more docs-friendly.

**Review by fowczarek — APPROVED** [review]

**Review by maarcingebala — APPROVED** [review]

**Review by eCzerniak — CHANGES_REQUESTED** [review]
image

`publishedFrom` and `availableFrom` are not working as they should. When I am using those filters then I get products whit different date then I choose.

**Comment by kadewu** [comment]
> image
> 
> `publishedFrom` and `availableFrom` are not working as they should. When I am using those filters then I get products whit different date then I choose.

Maybe I should change or better explain the description on those two filters. I believe that they work fine cause the idea is, to check if the `product` will be for example available in future. So if the date `availableAt` is smaller than what we asked for, then that means `product` is available.

**Review by eCzerniak — APPROVED** [review]

**Pull request merged**